pkg/cmd/cloudenv/options.go

```diff
@@ -0,0 +1,374 @@
+type Shell string
+
+const (
+	bash       Shell = "bash"
+	zsh        Shell = "zsh"
+	fish       Shell = "fish"
+	powershell Shell = "powershell"
+)
+
+var validShells = []Shell{bash, zsh, fish, powershell}
+
+// EvalCommand returns the script that evaluates the given command
+func (s Shell) EvalCommand(cmd string) string {
+	var format string
+
+	switch s {
+	case fish:
+		format = "eval (%s)"
+	case powershell:
+		// Invoke-Expression cannot execute multi-line functions!!!
+		format = "& %s | Invoke-Expression"
+	default:
+		format = "eval $(%s)"
+	}
+
+	return fmt.Sprintf(format, cmd)
+}
+
+// Prompt returns the typical prompt for a given os
+func (s Shell) Prompt(goos string) string {
+	switch s {
+	case powershell:
+		if goos == "windows" {
+			return "PS C:\\> "
+		}
+
+		return "PS /> "
+	default:
+		return "$ "
+	}
+}
+
+// AddCommand adds a shell sub command to a parent
+func (s Shell) AddCommand(parent *cobra.Command, runE func(cmd *cobra.Command, args []string) error) {
+	shortFormat := "generate the cloud provider CLI configuration script for %s"
+	longFormat := `Generate the cloud provider CLI configuration script for %s.
+
+To load the cloud provider CLI configuration script in your current shell session:
+%s
+`
+	cmdWithPrompt := s.Prompt(runtime.GOOS) + s.EvalCommand(fmt.Sprintf("%s %s", parent.CommandPath(), s))
+	shell := string(s)
+	cmd := &cobra.Command{
+		Use:   shell,
+		Short: fmt.Sprintf(shortFormat, shell),
+		Long:  fmt.Sprintf(longFormat, shell, cmdWithPrompt),
+		RunE:  runE,
+	}
+	parent.AddCommand(cmd)
+}
+
+// Validate checks if the shell is valid
+func (s Shell) Validate() error {
+	for _, shell := range validShells {
+		if s == shell {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("invalid shell given, must be one of %v", validShells)
+}
+
+// CloudProvider represent the type of cloud provider
+type CloudProvider string
+
+const (
+	alicloud  CloudProvider = "alicloud"
+	gcp       CloudProvider = "gcp"
+	openstack CloudProvider = "openstack"
+)
+
+// CLI returns the CLI for the cloud provider
+func (c CloudProvider) CLI() string {
+	switch c {
+	case alicloud:
+		return "aliyun"
+	case gcp:
+		return "gcloud"
+	default:
+		return string(c)
+	}
+}
+
+func getKeyStoneURL(cloudProfile *gardencorev1beta1.CloudProfile, region string) (string, error) {
+	config, err := gardenclient.CloudProfile(*cloudProfile).GetOpenstackProviderConfig()
+	if err != nil {
+		return "", fmt.Errorf("invalid providerConfig in CloudProfile %q", cloudProfile.Name)
+	}
+
+	if config.KeyStoneURL != "" {
+		return config.KeyStoneURL, nil
+	}
+
+	for _, keyStoneURL := range config.KeyStoneURLs {
+		if keyStoneURL.Region == region {
+			return keyStoneURL.URL, nil
+		}
+	}
+
+	return "", fmt.Errorf("cannot find keyStoneURL for region %q in CloudProfile %q", region, cloudProfile.Name)
+}
+
+func parseCredentials(secret *corev1.Secret, credentials interface{}) ([]byte, error) {
+	privateKey := secret.Data["serviceaccount.json"]
```

Comment:
Because it is not the `serviceAccount` but the private key used for authentication of this `serviceAccount` 
```sh
$ gcloud auth activate-service-account foo@example.org --key-file=/path/key.json --project=PROJECT_ID
```
To avoid confusion I renamed it to `data`.